Various cleanups of Handler.insertHandler

gregw opened the pull request:
Fixes #9297

gregw committed:
Various cleanups of Handler.insertHandler

gregw committed:
Various cleanups of Handler.insertHandler

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -1660,26 +1660,7 @@ else if (handler instanceof SecurityHandler)
-            doSetHandler(h, handler);
-            doSetHandler(tail, next);
-        }
+            super.insertHandler(handler);
```
The code is now very different. Is that ok?

`ServletContextHandlerTest.testReplaceHandler()` will fail if you add session and or security (not sure if failed before).

gregw replied:
@sbordet Well I couldn't work out what the existing code was doing...     So yeah it is different.   Happy to punt this to 12.0.4 so we don't break anything this release cycle.

joakime reviewed (changes requested):
When you run the EE10 `ServletContextHandler.testReplaceHandler()` in this branch you get a new WARN  logging event. (should probably rename this specific test case, as we don't really do "replace" anymore in that test case)

`2023-10-26 10:56:49.405:WARN :oejes.ServletContextHandler:main: ServletContextHandler.setHandler should not be called directly. Use insertHandler or setSessionHandler etc.`

Looks like `Handler.insertHandler()` winds up calling `ServletContextHandler.setHandler()` which triggers this warning.

That's a bit of a confusing warning.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ContextHandler.java:
```diff
@@ -2607,6 +2607,7 @@ protected void doStop() throws Exception
         @Override
         public void insertHandler(Singleton handler)
         {
+            // reimplemented insertHandler so we can directly call super.setHandler
```
Supposed to be a TODO? Otherwise it's unclear what the comment means.

gregw replied:
`o.e.j.ee9.nested.ContextHandler.setHandler` is not meant to be called directly and gives a warning that it should not be used.   So this code is exactly the super.insertHander code, but with `super.setHandler` called instead.

gregw replied:
replaced with:
```java
            // We cannot call super.insertHandler here, because it uses this.setHandler
            // which gives a warning.  This is the same code, but uses super.setHandler
```

sbordet commented:
Also, `ServletContextHandlerTest.testReplaceHandler()` fails if the `ServletContextHandler` in that test is configured with sessions or security.

What's the right semantic for `ServletContextHandler.insertHandler()`? 

Given `SCH -- [SessionH -- SecurityH -- ServletH]`, then `insertHandler()` of `H -- X -- Y` should result in `SCH -- H -- X -- Y -- [SessionH -- SecurityH -- ServletH]`?

And if you `insertHandler()` again with `Z`, is it after `Y`, or before `H`?

If that's the case, can you write that example in the javadocs of the overridden method?

gregw committed:
Various cleanups of Handler.insertHandler

gregw commented:
@sbordet @joakime I can't get that failure locally?  I've made some changes, but it is still a bit strange.  I've punted to the next release cycle.

gregw committed:
Various cleanups of Handler.insertHandler

gregw committed:
Various cleanups of Handler.insertHandler

gregw commented:
@sbordet @joakime nudge

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Handler.java:
```diff
@@ -335,11 +336,7 @@ default List<Handler> getHandlers()
-            {
-                tail = (Wrapper)tail.getHandler();
-            }
+            Singleton tail = Objects.requireNonNull(handler).getTail();
```
Remove unnecessary `Objects.requireNonNull()`, adds nothing to a naked call to `handler.getTail()`.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -1661,24 +1661,13 @@ else if (handler instanceof ServletHandler)
-            Singleton tail = handler.getTail();
+            // We cannot call super.insertHandler here, because it uses this.setHandler
+            // which gives a warning.  This is the same code, but uses super.setHandler
+            Singleton tail = Objects.requireNonNull(handler).getTail();
```
Remove `Objects.requireNonNull()`, unnecessary.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -1661,24 +1661,13 @@ else if (handler instanceof ServletHandler)
         {
-            Singleton tail = handler.getTail();
+            // We cannot call super.insertHandler here, because it uses this.setHandler
+            // which gives a warning.  This is the same code, but uses super.setHandler
```
Why does `ServletContextHandler.setHandler()` gives a warning?
Seems natural to me that it would set the "child" Handler, which for this class is after the `ServletHandler`, which is always present.

gregw replied:
@sbordet Because an ee9 nested ContextHandler (old API)  contains a core ContextHandler(new API) and sets up a little private chain:  coreContextHandler->core2nestedHandler->nestedContextHandler.
If we allow a setHandler(coreHandler) on core ContextHandler, then it breaks this chain.  We can insert core Handlers between coreContextHandler and core2nestedHandler, but we cannot replace the core2nestHandler without breaking the class.
We can however do a setHandler(nestedHandler) on the nested ContextHandler.

gregw replied:
Trying new semantic as suggested by @sbordet that setHandler does a set on the ServletHandler

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/ServletContextHandlerTest.java:
```diff
@@ -1900,15 +1901,26 @@ public void contextDestroyed(ServletContextEvent sce)
+                return false;
+            }
+        };
+        servletContextHandler.getTail().setHandler(last);
```
See above, this line would more naturally be `servletContextHandler.setHandler(last)`.

gregw replied:
@sbordet  ditto - that has the effect of replacing the sessionHandler/securityHandler/servletHandler that this class has provided. inserting handlers makes sense, but replacing breaks the class.   This is a warning we've had in jetty-9/10/11 for sometime.

I could replace it with:
```suggestion
        servletContextHandler.getServletHandler().setHandler(last);
```

But that assumes there is not already a handler set on the servlet handler.  This is more robust way of doing it

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ContextHandler.java:
```diff
@@ -2607,6 +2616,8 @@ protected void doStop() throws Exception
         public void insertHandler(Singleton handler)
         {
+            // We cannot call super.insertHandler here, because it uses this.setHandler
+            // which gives a warning.  This is the same code, but uses super.setHandler
```
See above about not issuing a warning, but defaulting to setting the "tail".

Also, why `setHandler()` below is overridden to throw?

gregw replied:
@sbordet Because setting the handler of the coreContextHandler breaks this class as it removes the core2nestedHandler.      Inserting is OK, but replacing is not.

sbordet commented on jetty-ee9/jetty-ee9-servlet/src/test/java/org/eclipse/jetty/ee9/servlet/ServletContextHandlerTest.java:
```diff
@@ -1852,14 +1853,23 @@ public void contextDestroyed(ServletContextEvent sce)
+
+        servletContextHandler.insertHandler(coreResourceHandler);
+        servletContextHandler.insertHandler(nestedHandler);
+        servletContextHandler.getTail().setHandler(last);
```
Calling `servletContextHandler.setHandler(last)` seems more natural.

gregw replied:
@sbordet  With what semantics?   Is that replacing any optional SessionHandler or SecurityHandler that may be included?  Does that replace the ServletHandler that should always be there?   or is it a set on the ServletHandler?  If it is called twice is it an insert on the ServletHandler?

We have always had that a handler.setHandler(next), sets the next of that handler and not some tail or chain etc.     In a few cases we disable setHandler when it is not appropriate.    I'm not sure I like defining new semantics

gregw replied:
done.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/9297/insertHandler

gregw committed:
Various cleanups of Handler.insertHandler

gregw commented:
@sbordet any chance of a re-review before you go on vacation, now that I've explained how core 2 nested Handlers work?

gregw committed:
Various cleanups of Handler.insertHandler

sbordet reviewed (changes requested):
@gregw I am still a bit confused by SCH interaction with `setHandler()` and `insertHandler()`.

I would like the following:
* Make `ServletHandler extends Handler.Abstract` so that it's clear that it has no child. It still can return `false`, only in particular cases where users configure it to *not* have a `DefaultServlet` (in which case we just return the "core" 404).
* `SCH.setHandler()` always sets the first child after SCH (like all `Singleton` do), overwriting the existing one, with the special handling for "the triplet", i.e. `Session|Security|ServletHandler`, and relinks the triplet.
* The triplet is always last in the chain, so `SCH.getTail()` always returns the `ServletHandler`.

Some example:

```
sch.setHandler(gzipH);
sch.insertHandler(wsUpgradeH); // => sch-wsuh-gziph-[sessionh]-[securityh]-servleth
sch.getHandler() == wsuh
sch.getTail() == servleth
```

With the current PR, the code above sets up the following chain, which I think is wrong: `sch-wsuh-[sessionh]-[securityh]-servleth-gziph`, as we would never gzip.

With the proposal above:
```
sch.setHandler(gziph);
sch.setHandler(wsuh); // sch-wsu-[sessionh]-[securityh]-servleth -- gziph is lost as expected.

sch.setHandler(B);
sch.getHandler() == B; // The current PR breaks this.
```

I am sure the proposal above is a change to what I said earlier, but I find it clearer.
Would it work for you?

joakime commented:
A quick test of this PR and it's handler tree looks like this ...

``` java
ServletContextHandler servletContextHandler = new ServletContextHandler(ServletContextHandler.SECURITY);

ServletHolder sh = new ServletHolder(new TestServlet());
servletContextHandler.addServlet(sh, "/foo");

servletContextHandler.setHandler(new GzipHandler());
servletContextHandler.insertHandler(new XYHandler());

Handler handler = servletContextHandler.getHandler(); // what does this look like ?
```

This results in the Handler tree ...

```
XYHandler 
   .handler = ConstraintSecurityHandler
      .handler = ServletHandler
          .handler = GzipHandler
```

Seems odd to have GzipHandler under the ServletHandler like this.
Would have expected it before the ConstraintSecurityHandler.

gregw commented:
@sbordet there are definitely usages of ServletHandler as a wrapper. We might even start putting a resource handler there.

We have survived decades with a bit of strangeness with SCH containing a triple chain.  I think you are trying to solve a problem that we don't have

gregw commented:
@joakime the following code:
```java
        Server server = new Server();
        ServletContextHandler servletContextHandler = new ServletContextHandler(ServletContextHandler.SECURITY);
        servletContextHandler.setHandler(new GzipHandler());
        servletContextHandler.insertHandler(new Handler.Wrapper());

        server.setHandler(servletContextHandler);
        server.start();
        servletContextHandler.dumpStdErr();
```

produces:
```
oeje10s.ServletContextHandler@740fb309{ROOT,/,b=null,a=AVAILABLE,h=oejs.Handler$Wrapper@7bd7d6d6{STARTED}} - STARTED
+= oejs.Handler$Wrapper@7bd7d6d6{STARTED} - STARTED
|  += oeje10ss.ConstraintSecurityHandler@28cda624{STARTED} - STARTED
|     += oeje10s.ServletHandler@4dc27487{STARTED} - STARTED
|     |  += GzipHandler@6a4f1a55{STARTED,min=32,inflate=-1} - STARTED
```

I think the insert behaviour is exactly what we want.
I agree it is a little strange that SCH.setHandler does some magic so that the GzipHandler is after the ServletHandler, but that is what @sbordet asked for.   I'm a 60:40 for the previous behaviour where setHandler was always a set on the current handler with minimal magic.    But there always was a little bit of magic and sometimes warnings.... so this approach just embraces the magic and avoids the warnings. 

Either way, I don't think it is hugely important, as I can't see I real use case for wanting to do setHandler in this situation.  However, with EE10 and beyond, it is slightly more important as users can no longer use a ServletHandler with a plain ContextHandler.  It only works with a ServletContextHandler (to avoid the need for a ScopedHandler mechanism we establish servlet scope as we enter the context).

As I said, I'm 60:40... but as I've currently implemented/tested  the 40... I'm happy to go with it if there is consensus???

gregw commented:
@sbortdet

> @gregw I am still a bit confused by SCH interaction with `setHandler()` and `insertHandler()`.

I think we have always had some strangeness here where one handle carries a tuple of other handlers that must be in a specific structure.   I'm not sure we can avoid t he situation.  We can only address confusion by a bit better consistency and better documentation, which is what I think this PR does.  We can't make the issue go away.


> I would like the following:
> 
> * Make `ServletHandler extends Handler.Abstract` so that it's clear that it has no child. It still can return `false`, only in particular cases where users configure it to _not_ have a `DefaultServlet` (in which case we just return the "core" 404).

@sbordet there are definitely usages of `ServletHandler` as a wrapper (see below).  This is not an option.

> * `SCH.setHandler()` always sets the first child after SCH (like all `Singleton` do), overwriting the existing one, with the special handling for "the triplet", i.e. `Session|Security|ServletHandler`, and relinks the triplet.

That's what I've now implemented for you.   If it is not a `Session|Security|ServletHandler`, it always sets the handler of the `ServletHandler`, overwriting the existing one.  So it is setting the first child after SCH.

> * The triplet is always last in the chain, so `SCH.getTail()` always returns the `ServletHandler`.

Not an option.  There are valid use-cases for falling through a `ServletHandler`.   For example we can put a `ResourceHandler` after the `ServletHandler` and not have a `DefaultServlet`(with all it's complexities).  



> Some example:
> 
> ```
> sch.setHandler(gzipH);
> sch.insertHandler(wsUpgradeH); // => sch-wsuh-gziph-[sessionh]-[securityh]-servleth
> sch.getHandler() == wsuh
> sch.getTail() == servleth
> ```
> 
> With the current PR, the code above sets up the following chain, which I think is wrong: `sch-wsuh-[sessionh]-[securityh]-servleth-gziph`, as we would never gzip.

I said I was 60:40 in my reply to @joakime, but now I'm 95:5 for the former behaviour of setHandler with less magic since the following should always be true:
```
    sch.setHandler(xyzHandler);
    sch.getHandler() == xyzHandler
```

I think the less magic, then the less confusion.    My intention is to go back to a simple setHandler with warnings for bad behaviours.    Perhaps it should not even intercept `Session|Security|ServletHandler`

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/9297/insertHandler

# Conflicts:
#	jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/SessionHandler.java

gregw committed:
Various cleanups of Handler.insertHandler

gregw committed:
Various cleanups of Handler.insertHandler

gregw commented:
@joakime nudge

gregw commented:
@sbordet nudge

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/9297/insertHandler

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -1647,6 +1638,22 @@ public void setServletHandler(ServletHandler servletHandler)
+    @Override
+    public void setHandler(Handler handler)
+    {
+        if (handler instanceof SessionHandler && getHandler() instanceof SessionHandler)
+            setSessionHandler((SessionHandler)handler);
+        else if (handler instanceof SecurityHandler && getHandler() instanceof SecurityHandler)
+            setSecurityHandler((SecurityHandler)handler);
+        else if (handler instanceof ServletHandler && getHandler() instanceof ServletHandler)
+            setServletHandler((ServletHandler)handler);
```
I don't think this logic is right.
Calling `setHandler()` with one of the 3 handlers will never work.

gregw replied:
It has been this way for several years at least.  So we are not making it special. It is already special.
I'm not sure we want to put a breaking change into the API without thought.   I found a few places in the code that did rely on this behaviour, so other users may also.

I think if we want to change setHandler, that should be a different PR with more consideration.

gregw replied:
@sbordet I've put the implementation of setHandler back to exactly what it has been for some time.  However, this PR removes any usage of it that depends on that behaviour.    If it is an issue, then we can make the method always warn if called on SCH.

sbordet reviewed (changes requested):
I am still confused.

I think we should stick to the behavior of the other `Handler`s and not make this one any special.

As I said, I was wrong wanting `setHandler()` to set the child of `ServletHandler`.
I think it should just do what any other `Handler` does, i.e. setting the immediate child.

`insertHandler()` would have the same semantic as others, i.e. setting the immediate child and shifting the current child.

The "triplet" always come last.

I'm fine for `ServletHandler` to be a `Wrapper` if there are use cases, but setting its child should always be done explicitly, i.e. `sch.getServletHandler().setHandler(...)`, not via some "magic" in `SCH.setHandler()`.

gregw committed:
Various cleanups of Handler.insertHandler

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/9297/insertHandler

sbordet committed:
Added missing call to relinkHandlers() in setHandler() after calling super.
Moved call to relinkHandlers() in insertHandler(), as the various setSession|Security|ServletHandler() already call relinkHandlers().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet reviewed (approved):
@gregw see my last commit where I added in one place and moved in another the call to relinkHandlers().
LGTM with the current behavior.

gregw commented:
> @gregw see my last commit where I added in one place and moved in another the call to relinkHandlers(). LGTM with the current behavior.

@sbordet 

The move of the relink in `insertHandler` is good, as it saves doing it twice.

But the `relinkHandlers` in `setHandler` is wrong (or perhaps the `relinkHandlers` impl is not good enough.       If I try to do a set chain of handlers A->B->C->SessionHandler->ServletHandler, then relinkHandlers will look at A, see that its next handler is not SessionHandler and replace it, so the chain will end up A->SessionHandler->ServletHandler and B->C will be dropped.

I'll make a unit test to check this and then try to fix relinkHandler.... but I just don't think we should try to do magic with anything other than the Session/Security/Servlet handlers.  Even that magic is probably too much.

gregw commented:
@sbordet OK I was wrong.  `relinkHandlers` is good enough to follow the chain.   But it is just more magic.  I don't like it, but OK.
